webgoat-lessons/challenge/src/main/java/org/owasp/webgoat/challenges/challenge7/MD5.java

```diff
@@ -0,0 +1,689 @@
+            0, 0, 0, 0, 0, 0, 0, 0,
+            0, 0, 0, 0, 0, 0, 0, 0,
+            0, 0, 0, 0, 0, 0, 0, 0,
+            0, 0, 0, 0, 0, 0, 0, 0,
+    };
+
+    /**
+     * Contains internal state of the MD5 class.
+     * Passes MD5 test suite as defined in RFC1321.
+     *
+     * @since ostermillerutils 1.00.00
+     */
+    private class MD5State {
+
+        /**
+         * True if this state is valid.
+         *
+         * @since ostermillerutils 1.00.00
+         */
+        private boolean valid = true;
+
+        /**
+         * Reset to initial state.
+         *
+         * @since ostermillerutils 1.00.00
+         */
+        private void reset() {
+            state[0] = 0x67452301;
+            state[1] = 0xefcdab89;
+            state[2] = 0x98badcfe;
+            state[3] = 0x10325476;
+
+            bitCount = 0;
+        }
+
+        /**
+         * 128-byte state
+         *
+         * @since ostermillerutils 1.00.00
+         */
+        private int state[] = new int[4];
+
+        /**
+         * 64-bit count of the number of bits that have been hashed.
+         *
+         * @since ostermillerutils 1.00.00
+         */
+        private long bitCount;
+
+        /**
+         * 64-byte buffer (512 bits) for storing to-be-hashed characters
+         *
+         * @since ostermillerutils 1.00.00
+         */
+        private byte buffer[] = new byte[64];
+
+        private MD5State() {
+            reset();
+        }
+
+        /**
+         * Set this state to be exactly the same as some other.
+         *
+         * @param from state to copy from.
+         * @since ostermillerutils 1.00.00
+         */
+        private void copy(MD5State from) {
+            System.arraycopy(from.buffer, 0, this.buffer, 0, this.buffer.length);
+            System.arraycopy(from.state, 0, this.state, 0, this.state.length);
+            this.valid = from.valid;
+            this.bitCount = from.bitCount;
+        }
+    }
+
+
+    /**
+     * Turns array of bytes into string representing each byte as
+     * a two digit unsigned hex number.
+     *
+     * @param hash Array of bytes to convert to hex-string
+     * @return Generated hex string
+     * @since ostermillerutils 1.00.00
+     */
+    private static String toHex(byte hash[]) {
+        StringBuffer buf = new StringBuffer(hash.length * 2);
+        for (byte element : hash) {
+            int intVal = element & 0xff;
+            if (intVal < 0x10) {
+                // append a zero before a one digit hex
+                // number to make it two digits.
+                buf.append("0");
+            }
+            buf.append(Integer.toHexString(intVal));
+        }
+        return buf.toString();
+    }
+
+    private static int FF(int a, int b, int c, int d, int x, int s, int ac) {
+        a += ((b & c) | (~b & d));
+        a += x;
+        a += ac;
+        //return rotateLeft(a, s) + b;
+        a = (a << s) | (a >>> (32 - s));
+        return a + b;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 533:538, 542:547.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.